lib/Dialect/HW/HWTypes.cpp

```diff
@@ -108,40 +108,16 @@ bool circt::hw::isHWValueType(Type type) {
 /// value of this type. Returns -1 if the type is not known or cannot be
 /// statically computed.
 int64_t circt::hw::getBitWidth(mlir::Type type) {
+  // First, check if the type implements the BitWidthTypeInterface.
+  if (auto iface = dyn_cast<BitWidthTypeInterface>(type)) {
+    auto width = iface.getBitWidth();
+    return width.has_value() ? *width : -1;
+  }
+
+  // Handle built-in types that don't implement the interface.
   return llvm::TypeSwitch<::mlir::Type, int64_t>(type)
       .Case<IntegerType>(
           [](IntegerType t) { return t.getIntOrFloatBitWidth(); })
```

Comment:
nit: Can we check integer type first and later down cast to BitWdithTypeInterface? It might be unnecessary micro optimization but hw::getBitWidth is called tremendous time in verifier etc so I think it's worth micro optimize. 

FYI down casting to interface has few more cost than down casting to concrete type.  
https://youtu.be/7qvVMUSxqz4?si=otqaKjiCq7lqF-qR&t=1088